keps/407-gang-scheduling/README.md

```diff
@@ -0,0 +1,407 @@
+the scheduler may prioritize scheduling leader pods while leaving their worker pods pending. 
+This results in a situation where no inference service can serve properly, as each service requires both leader and worker pods to be running.
+Gang scheduling support would ensure that pods within a replica are either all scheduled together or not scheduled at all, 
+thus preventing resource deadlocks and maintaining service availability.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Create different PodGroup resources for each lws replica for different custom schedulers (Volcano, scheduler-plugins, YuniKorn etc.) to enable gang scheduling
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Implement specific gang scheduling policies
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+LeaderWorkerSet will create a PodGroup resource for each replica and handle the association between pods and PodGroup. When creating or updating a LeaderWorkerSet, the controller will:
+
+1. Create PodGroup resources for each replica
+2. Ensure proper association between pods and their corresponding PodGroup
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user running distributed inference services with LeaderWorkerSet, I want to prevent resource deadlocks and ensure service availability in resource-constrained environments.
+
+Currently, when cluster resources are limited, the scheduler might only schedule leader pods while leaving worker pods pending. This results in inference services being unavailable despite consuming cluster resources.
+
+With gang scheduling support, leader and worker pods will be treated as a single scheduling unit, ensuring they are either all scheduled together or remain pending as a group. This prevents resource waste and ensures resources are only used by fully functional inference services.
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
```

Comment:
There are cases when a pod needs to be re-created for a group, or a group needs to scale (and create new pods). How is this handled for the group? You can imagine a case where the operation is not possible (or would take a long time in the queue of pending pods) and you get something akin to blocking.

Reply:
Gotcha, so you can resize (and create a new set of workers or leader to generate a new group) but an existing set cannot dynamically scale.

Reply:
Yes, this is also designed based on the current features of lws. Otherwise, I would directly create PodGroups in the lws controller, but this would require a lot of extra handling and be more complex